src/assets.zig

```diff
@@ -294,29 +297,65 @@ fn registerRecipesFromZon(zon: ZonElement) void {
+			.object => try loadFromZonLegacy(allocator, zon),
+			.array => try loadFromZon(allocator, zon),
+			else => return error.InvalidPaletteFormat,
+		};
+		std.debug.assert(firstElement == null or std.mem.eql(u8, self.palette.items[0], firstElement.?));
+		return self;
+	}
+	fn initEmpty(allocator: NeverFailingAllocator, firstElement: ?[]const u8) !*@This() {
+		const self = allocator.create(Palette);
+		self.* = Palette{
+			.palette = .initCapacity(allocator, 128),
+		};
+		if(firstElement) |elem| self.palette.append(allocator.dupe(u8, elem));
+		return self;
+	}
+	fn loadFromZon(allocator: NeverFailingAllocator, zon: ZonElement) !*@This() {
 		const self = allocator.create(Palette);
 		self.* = Palette{
-			.palette = .init(allocator),
+			.palette = .initCapacity(allocator, zon.array.items.len),
 		};
 		errdefer self.deinit();
-		if(zon != .object or zon.object.count() == 0) {
-			if(firstElement) |elem| self.palette.append(allocator.dupe(u8, elem));
-		} else {
-			const palette = main.stackAllocator.alloc(?[]const u8, zon.object.count());
-			defer main.stackAllocator.free(palette);
-			for(palette) |*val| {
-				val.* = null;
-			}
-			var iterator = zon.object.iterator();
-			while(iterator.next()) |entry| {
-				palette[entry.value_ptr.as(usize, std.math.maxInt(usize))] = entry.key_ptr.*;
-			}
-			if(firstElement) |elem| std.debug.assert(std.mem.eql(u8, palette[0].?, elem));
-			for(palette) |val| {
-				std.log.info("palette[{}]: {s}", .{self.palette.items.len, val.?});
-				self.palette.append(allocator.dupe(u8, val orelse return error.MissingKeyInPalette));
+
+		for(zon.array.items) |name| {
+			const stringId = name.as(?[]const u8, null) orelse return error.InvalidPaletteFormat;
+			self.palette.appendAssumeCapacity(allocator.dupe(u8, stringId));
+		}
+		return self;
+	}
+	fn loadFromZonLegacy(allocator: NeverFailingAllocator, zon: ZonElement) !*@This() {
+		// Using zon.object.count() here has the implication that array can not be sparse.
+		const paletteLength = zon.object.count();
+		const translationPalette = main.stackAllocator.alloc([]const u8, paletteLength);
+		defer main.stackAllocator.free(translationPalette);
+
+		var iterator = zon.object.iterator();
+		while(iterator.next()) |entry| {
+			const numericId = entry.value_ptr.as(?usize, null) orelse return error.InvalidPaletteFormat;
+			const name = entry.key_ptr.*;
+
+			if(numericId >= translationPalette.len) {
+				std.log.err("ID {} ('{s}') out of range. This can be caused by palette having missing block IDs.", .{numericId, name});
+				return error.SparsePaletteNotAllowed;
 			}
+			translationPalette[numericId] = name;
+		}
+
+		const self = allocator.create(Palette);
+		self.* = Palette{
+			.palette = .initCapacity(allocator, paletteLength),
+		};
+		errdefer self.deinit();
+
+		for(translationPalette) |val| {
+			std.log.info("palette[{}]: {s}", .{self.palette.items.len, val});
+			self.palette.appendAssumeCapacity(allocator.dupe(u8, val));
```

Comment:
I would prefer it if you didn't reformat the legacy version as much to make reviewing easier.
It will be removed in #480 anyways.

Reply:
Boy scout rule is strong in me.